### ISSUE TITLE
ConvertTo-Json does not convert additional members of nested properties

### ISSUE THREAD

**Issue by JTDotNet** [PowerShell/PowerShell#19185]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://docs.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I am attempting to add a NoteProperty to a reference type property of an object, then convert that entire object to JSON using the ConvertTo-Json cmdlet. While this works fine when adding a NoteProperty to a top level object, if you do this to an object that's a property of the top-level object, the NoteProperty is not handled by the cmdlet:

`$foo = Get-ADUser "someUser"

$foo.SID | Add-Member -MemberType NoteProperty -Name "bar" -Value "barValue" -Force

$foo | ConvertTo-Json`

### Expected behavior

```console
Cmdlet should output:

{
  ...
  "SID": {
    "BinaryLength": 28,
    "AccountDomainSid": {
      "BinaryLength": 24,
      "AccountDomainSid": [redacted],
      "Value": [redacted]
    },
    "Value": [redacted],
    "bar": "barValue"
  },
  ...
}
```

### Actual behavior

```console
Output does not include additional property:

{
  ...
  "SID": {
    "BinaryLength": 28,
    "AccountDomainSid": {
      "BinaryLength": 24,
      "AccountDomainSid": [redacted],
      "Value": [redacted]
    },
    "Value": [redacted]
  },
  ...
}
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.3.2
PSEdition                      Core
GitCommitId                    7.3.2
OS                             Microsoft Windows 10.0.19044
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by mklement0** [comment]
I think the problem is specific to type of objects output by `Get-ADUser`, as the behavior with other types is as expected; e.g.:

```powershell
[pscustomobject] @{
  Foo = [regex] '...' | Add-Member -PassThru -NotePropertyName Bar -NotePropertyValue 'Bar!'
} | ConvertTo-Json -Depth 1 3>$null
```

Result (as expected; note the presence of `"Bar"`):
```json
{
  "Foo": {
    "Options": 0,
    "RightToLeft": false,
    "MatchTimeout": "-00:00:00.0010000",
    "Bar": "Bar!"
  }
}
```

I _suspect_:

* the problem is related to `Get-ADUser` output objects being _dynamic_ objects.
* you may be able to _work around_ the problem by creating a `[pscustomobject]` clone of the AD user first:

```powershell
# Untested; note the "Select-Object -Property *"
$foo = Get-ADUser "someUser" | Select-Object -Property *
$foo.SID | Add-Member -MemberType NoteProperty -Name "bar" -Value "barValue" -Force
$foo | ConvertTo-Json

**Comment by mklement0** [comment]
It does seem like a bug that affects `ConvertTo-Json`, specifically, and it is connected to how the NoteProperty is added to the object property.

Using a simplified example:

```powershell
# Sample object.
$o = [pscustomobject] @{ Foo = [regex] '...' }

# Directly decorate a property of the object with a NoteProperty.
$o.Foo | Add-Member -NotePropertyName Bar -NotePropertyValue 'Bar!'

# !! "Bar" is missing,
# !! even though it does show when you output $o.Foo alone (it is reflected in $o.Foo.psobject.PSProperties)
$o | ConvertTo-Json -Depth 1 3>$null
```

_Workaround_: 

```powershell
# Sample object.
$o = [pscustomobject] @{ Foo = [regex] '...' }

# WORKAROUND: Use -PassThru to output the decorated property value and re-assign.
$o.Foo = $o.Foo  | Add-Member -PassThru -NotePropertyName Bar -NotePropertyValue 'Bar!'

# OK - now "Bar" shows up.
$o | ConvertTo-Json -Depth 1 3>$null
```

That this makes a difference appears to be connected to the fact that the object stored in `.Foo` then has an invisible `[psobject]` wrapper.

In fact, in the first snippet, if you insert `$o.Foo = [psobject] $o.Foo` after the `Add-Member` call, you get the same effect.

**Comment by mklement0** [comment]
Without having looked a the source code, I think the bug comes down to this (as mostly implied by your original report):

`ConvertTo-Json` honors ETS instance properties:

* _always_ on the _input objects themselves_.

* but when it walks each input object's object graph, it honors them _ONLY if the property values happen to be `[psobject]`-wrapped_ (such as would happen if the value is the result of a cmdlet call). 
  * This fact can be used as a _workaround_ for now, by re-assigning a property's value to itself with a `[psobject]` cast (e.g., `$obj.Prop = [psobject] $obj.Prop`

A simplified example:

```powershell
$prop = [regex] '...'
$prop | Add-Member PropNoteProp PropNoteProp!

$o = [pscustomobject] @{
  PropUnwrapped = $prop
  PropWrapped = [psobject] $prop
}

$o | Add-Member ObjNoteProp ObjNoteProp!

$o | ConvertTo-Json -Depth 1 3>$null
```

Output (the `"PropUnwrapped"` value is missing the `"PropNoteProp"` property):

```json
{
  "PropUnwrapped": {
    "Options": 0,
    "RightToLeft": false,
    "MatchTimeout": "-00:00:00.0010000"
  },
  "PropWrapped": {
    "Options": 0,
    "RightToLeft": false,
    "MatchTimeout": "-00:00:00.0010000",
    "PropNoteProp": "PropNoteProp!"
  },
  "ObjNoteProp": "ObjNoteProp!"
}
```

**Comment by JTUoB** [comment]
@mklement0 thanks for the reply and explanation - that makes sense. Unfortunately I haven't been able to get it to work in my case. It looks like the object I'm working with does not get wrapped when assigned an explicitly typed psobject variable - the below returns false for the -is [psobject] line:

`$gatewayClusters = Get-DataGatewayCluster

foreach ($cluster in $gatewayClusters) {
    for ($i = 0; $i -lt $cluster.MemberGateways.Count; $i++){

        $gateway = $cluster.MemberGateways[$i]

        $annotations = $gateway.Annotation | ConvertFrom-Json -AsHashtable -Depth 100
        foreach ($annotation in $annotations.GetEnumerator()) {
            $gateway | Add-Member -MemberType NoteProperty -Name $annotation.Name -Value $annotation.Value -Force
        }

        $cluster.MemberGateways[$i] = [psobject]$gateway
        $cluster.MemberGateways[$i] -is [psobject]
    }
}`

FYI the type I'm trying to coerce is the undocumented Microsoft.PowerBI.ServiceContracts.Api.MemberGateway, which does have a setter (I just used Get-ADUser in the OP as I thought it would be better to use a more familiar example). I'm not sure why this type doesn't get coerced when regex does - are there some types that aren't compatible with psobject, or do types need to inherit something to be wrapped by psobject?

**Comment by mklement0** [comment]
Good to know, @JTUoB.

A _guess_ I had earlier was that perhaps _dynamic_ types - those implementing the [`IDynamicMetaObjectProvider`](https://learn.microsoft.com/en-US/dotnet/api/System.Dynamic.IDynamicMetaObjectProvider) interface  - interfere with ETS instance properties, and perhaps your `$gateway` variable contains an instance of such a type - 
`$gateway -is [System.Dynamic.IDynamicMetaObjectProvider]` would reveal that.

A possible workaround is to create a `[pscustomobject]` _clone_ of your object and decorate _it_:

```powershell
$gateway = $cluster.MemberGateways[$i] | Select-Object -Property *
```

**Comment by JTUoB** [comment]
Thanks @mklement0  for the suggestion - I did get false returned for is IDynamicMetaObjectProvider, but I suppose it may be possible that it's implementing another interface or inheriting from another type that causes problems. I wonder if it's because I'm attempting to assign a property of a .NET class, rather than a pscustomobject, and this just unwraps the object on assignment so that the property is stored as its actual type rather than psobject?

In any case, there's no way of me changing the behaviour of the MemberGateway object, so I'll just have to workaround it. I think the neatest solution, since I'm essentially just trying to deserialise a property with JSON text into persisted storage, is simply to deserialise it when it comes out of storage instead. 

It's good to learn a bit more about how PowerShell works under the hood. From what you've told me, it sounds like everything is working as designed, so I'll close this ticket down. Thanks again for your help with this, it's much appreciated.

**Comment by mklement0** [comment]
Glad to hear my comments were helpful, @JTUoB.

The tl;dr is: **I think you should reopen this issue**; see the bottom section for a summary.

---

> I wonder if it's because I'm attempting to assign a property of a .NET class, rather than a pscustomobject

Excellent point - I totally missed that.
You're right: the `[psobject]` wrapper is lost when assigning, _except_ when assigning to a `[pscustomobject]` instance's properties, so the workaround only works for `[pscustomobject]`.

(My `IDynamicMetaObjectProvider` hunch turns out to be mostly irrelevant to this discussion, though, for the record, the following - highly contrived - example, fails to serialize the ETS property even on the input object itself (and treats the dynamic object like a _dictionary_, which it also is):
```powershell
$do = [System.Dynamic.ExpandoObject]::new()  # ExpandoObject implements IDynamicMetaObjectProvider
$do.foo = 1; Add-Member -InputObject $do bar 2
ConvertTo-Json $do # !! Doesn't serialize "bar"
```
)

Note that, since v3, `[psobject]` wrappers are no longer needed for associating ETS instance members with arbitrary objects, so **the `[psobject]` cast, where even feasible, is truly just a _workaround_ that _shouldn't be necessary_**.
E.g., it works fine when accessing such instance members attached even to (native) _properties_ directly, albeit only with instance of .NET _reference_ types (other than `[string]`), as the following example demonstrates:

```powershell
# Create a sample .NET type.
Add-Type @'
  public class Foo {
    public System.Text.RegularExpressions.Regex Regex;
    public double Double;
    public string String;
  }
'@

# Instantiate the type and assign property values.
$o=[foo]::new()
$o.Regex = [regex] '.'
$o.Double = 42.0
$o.String = 'text'

# Try to decorate (attach ETS instance members to) the property values:

# OK: 
# Instances of .NET *reference* types (other than [string]).
$o.Regex | Add-Member foo 1

# !! NOT OK - QUIETLY IGNORED:
# !!  * Trying to decorate instance of .NET *value* types or [string] instances
# !!    that are PROPERTIES of a .NET Type other than [pscustomobject] does NOT work.
# !!  * You CAN, however, decorate STAND-ALONE INSTANCES of such types - see below.
# !!    That said, decorating instances of these types is BRITTLE and BEST AVOIDED.

# !! .NET *value* type:
$o.Double | Add-Member foo 2 # !! QUIETLY IGNORED
# !! As noted, decorating a *stand-alone* value-type instance works,
# !! presumably because PowerShell *boxes* the value and associates the ETS
# !! property with the boxed instance:
# !!   $dbl = 42.0; $dbl | Add-Member foo 2; $dbl.foo # OK
# !!   However, on binding the boxed instance to a [double]-typed 
# !!   parameter / constrained variable, it is *lost*; e.g., the following produces
# !!   NO output:
# !!      & { param([double] $dblParam) $dblParam.foo } $dbl # !! .foo is lost
# !!   If you removed the [double] typing (or used [object]), the boxed instance
# !!   *would* be passed, retaining the  .foo property

# !! [string] type:
$o.String = $o.String | Add-Member -PassThru foo 3 # !! QUIETLY IGNORED
# !! As noted, decorating a *stand-alone* [string] instance works, but only via
# !! [psobject] wrapper that must be obtained via the *return* value from
# !! Add-Member, via the -PassThru switch.
# !! When assigning to a native *property*, as shown, the wrapper is *lost*.
# !! Like decorated stand-alone value-type instance, stand-alone [string] instances
# !! lose their decoration when bound to a [string] parameter / constained variable,
# !! and also as the result of string *operations*:
# !!    $str = 'hi'; $str = $str | Add-Member -PassThru foo 3; $str.foo # OK
# !!    & { param([string] $strParam) $strParam.foo } $str # !! .foo is lost
# !!    ($str -replace 'x).foo # !! .foo is lost

# Show the results - only .Regex will have a .foo property.
$o | Format-Custom -Depth 1
```

---

**Summary**:

`ConvertTo-Json`'s behavior is _inconsistent_: if it honors ETS instance properties on the input objects themselves, it should also do so for their propeties.

  * Other cmdlets already do that, such as `Format-Custom` in the sample code above, as well as `Export-CliXml`

At the very least the limitation should be _documented_.

**Comment by JTUoB** [comment]
That's a good explanation. I agree it is inconsistent so will reopen.

**Comment by Shashtsoh** [comment]
Just encountered this behavior using PS7.
